packages/flutter_tools/lib/src/windows/visual_studio_solution_utils.dart

```diff
@@ -0,0 +1,318 @@
+const String _solutionTypeGuidVcxproj = '8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942';
+
+// The GUID for the folder above, managed by this class. This is an arbitrary
+// value that was randomly generated, but it should not be changed since that
+// would cause issues for existing Flutter projects.
+const String _flutterPluginSolutionFolderGuid = '5C2E738A-1DD3-445A-AAC8-EEB9648DD07C';
+// The FlutterBuild project GUID. This is an arbitrary
+// value that was randomly generated, but it should not be changed since that
+// would cause issues for existing Flutter projects.
+const String _flutterBuildProjectGuid = '6419BF13-6ECD-4CD2-9E85-E566A1F03F8F';
+
+/// Extracts and stores the plugin name and vcxproj GUID for [plugin].
+class _PluginProjectInfo {
+  _PluginProjectInfo(Plugin plugin) {
+    name = plugin.name;
+    final File projectFile = globals.fs.directory(plugin.path).childDirectory('windows').childFile('plugin.vcxproj');
+    guid = VisualStudioProject(projectFile).guid;
+    if (guid == null) {
+      throwToolExit('Unable to find a plugin.vcxproj for plugin "$name"');
+    }
+  }
+
+  // The name of the plugin, which is also the name of the symlink folder.
+  String name;
+
+  // The GUID of the plugin's project.
+  String guid;
+}
+
+/// Updates the solution file for [project] to have the project references and
+/// dependencies to include [plugins], removing any previous plugins from the
+/// solution.
+Future<void> updatePluginsInSolution(WindowsProject project, List<Plugin> plugins) async {
+  final String solutionContent = await project.solutionFile.readAsString();
+
+  // Map of GUID to name for the current plugin list.
+  final Map<String, String> currentPluginInfo = _getWindowsPluginNamesByGuid(plugins, project.pluginConfigKey);
+
+  // Find any plugins referenced in the project that are no longer used, and
+  // any that are new.
+  //
+  // While the simplest approach to updating the solution would be to remove all
+  // entries associated with plugins, and then add all the current plugins in
+  // one block, Visual Studio has its own (unknown, likely data-structure-hash
+  // based) order that it will use each time it writes out the file due to any
+  // solution-level changes made in the UI. To avoid thrashing, and resulting
+  // confusion (e.g., in review diffs), this update attempts to instead preserve
+  // the ordering that is already there, so that once Visual Studio has
+  // reordered the plugins, that order will be stable.
+  final Set<String> existingPlugins = _findPreviousPluginGuids(solutionContent);
+  final Set<String> currentPlugins = currentPluginInfo.keys.toSet();
+  final Set<String> removedPlugins = existingPlugins.difference(currentPlugins);
+  final Set<String> addedPlugins = currentPlugins.difference(existingPlugins);
+
+  final RegExp projectStartPattern = RegExp(r'^Project\("{' + _solutionTypeGuidVcxproj + r'}"\)\s*=\s*".*",\s*"(.*)",\s*"{([A-Fa-f0-9\-]*)}"\s*$');
```

Comment:
I don't have an alternative, but this series of regexps is a bit scary to me. Long term, it seems like it would be ideal if we had a Dart parser for the .sln format

Reply:
I'll file a bug before landing. I think a full parser will actually be more complex and have more regexes, which is part of why I decided not to go that way initially. The current approach definitely feels ugly, but the alternative is being able to understand every part of the solution file enough to re-output it correctly, even the parts we don't need to understand. Maybe there's a reasonable way to do that without looking too closely at the parts we don't need though?

Reply:
There are some existing Dart libraries to write parsers that are fairly high level, like https://github.com/petitparser/dart-petitparser (though performance is so-so). I agree, it would be more complex, but I think long term it would also be easier to maintain as it should separate the mechanics of parsing from how we intend to interact with the parsed data structure.

I don't have a better short-term solution, I think what you've got here is fine for now.

Reply:
I was thinking about it more last night, and probably if we are loose about tokenizing areas we don't understand (since I don't know that there's a formal description anywhere of everything that could be here, especially in the extension section), such as just taking whole lines without looking at them, then it probably wouldn't be too bad. I'll put some initial noted in a bug and add a link to it in a comment in the file.